### DISCUSSION TITLE
Parsing URL array-formatted parameter

### DISCUSSION THREAD

**Discussion by mergesort** [pointfreeco/swift-url-routing#107] [Q&A] [answered]
I'm looking to create a new route for my app, which will take a `urls` parameter which contains multiple comma-separated values, and parse them into domain-specific objects I can display. An example URL looks something like this:
```
https://plinky.app/links?urls=https://www.apple.com,https://www.google.com,https://www.microsoft.com
```

Presumably I can parse the parameter by separating out the comma-separated values like so:
```
let parsedURLs = urls.split(separator: ",").compactMap { URL(string: String($0)) }
```

But I was wondering if there was a built-in way/better to do this with `Query` and `Field`, that way I don't have to store this case's property as a `String`?

Thanks a lot!

**Comment by mbrandonw** [comment, marked as answer]
Hi @mergesort, the second argument of `Field` is a `Conversion`, which describes a way of transforming an input to an output and back into an input. It's similar to a parser, but it does not consume anything. It just transforms back-and-forth.

So you can create a custom conversion for turning a string into an array of `URL`s and back:

```swift
struct CommaSeparatedURLs: Conversion {
  func apply(_ input: Substring) throws -> [URL] {
    input.split(separator: ",").compactMap { URL(string: String($0)) }
  }
  func unapply(_ output: [URL]) throws -> Substring {
    output.map(\.absoluteString).joined(separator: ",")[...]
  }
}
```

And then you can use this with `Field`:

```swift
Field("links", CommaSeparatedURLs())
```

And you can also add a `static var` to `Conversion` to give a nicer syntax if you want.

**Reply by mergesort** [reply]
Thank you so much for the idea and the prompt reply @mbrandonw! I had taken a look before to see if I could recreate what y'all do with `.string`, `Digits()`, etc, but hadn't figured it out. This is so simple that I'm now going to apply this technique to a few other places I've resorted to saving string-representations. 

Unless I missed it I don't believe I saw this documented anywhere, if not it's definitely worth adding somewhere people can find. 😄

Thanks again!
